services/src/main/scala/cromwell/pipeline/service/ProjectSearchService.scala

```diff
@@ -0,0 +1,115 @@
+import scala.util.{ Failure, Success }
+
+trait ProjectSearchService {
+
+  def searchProjects(request: ProjectSearchRequest, userId: UserId): Future[Seq[Project]]
+
+}
+
+object ProjectSearchService {
+
+  object Exceptions {
+    sealed abstract class ProjectSearchException(message: String) extends ServiceException(message)
+    final case class InternalError(message: String = "Internal error") extends ProjectSearchException(message)
+  }
+
+  def apply(
+    projectService: ProjectService,
+    projectFileService: ProjectFileService,
+    projectConfigurationService: ProjectConfigurationService
+  )(
+    implicit ec: ExecutionContext
+  ): ProjectSearchService =
+    new ProjectSearchService {
+
+      override def searchProjects(request: ProjectSearchRequest, userId: UserId): Future[Seq[Project]] = {
+        val userProjects = projectService.getUserProjects(userId).recoverWith {
+          case _ => internalError("fetch projects")
+        }
+
+        def doSearch(searchFilter: ProjectSearchFilter): Future[Seq[Project]] = searchFilter match {
+          case AllFilter                    => userProjects
+          case ByNameFilter(mode, value)    => searchWithNameFilter(mode, value, userProjects)
+          case ByFilesFilter(_, hasFiles)   => searchByFilesFilter(hasFiles, userProjects, userId)
+          case ByConfigFilter(_, hasConfig) => searchByConfigFilter(hasConfig, userProjects, userId)
+          case ComplexOrFilter(leftFilter, rightFilter) =>
+            doSearch(leftFilter).flatMap(
+              resultLeft => doSearch(rightFilter).map(resultRight => resultLeft ++ resultRight)
+            )
+          case ComplexAndFilter(leftFilter, rightFilter) =>
+            doSearch(leftFilter).flatMap(
+              resultLeft => doSearch(rightFilter).map(resultRight => resultLeft.intersect(resultRight))
+            )
+        }
+        doSearch(request.filter).map(_.distinct)
+      }
+
+      private def searchByConfigFilter(
+        hasConfig: Boolean,
+        eventualProjects: Future[Seq[Project]],
+        userId: UserId
+      ): Future[Seq[Project]] = {
+        def getProjectWithConfig(project: Project): Future[(Project, Option[ProjectConfiguration])] =
+          projectConfigurationService.getLastByProjectId(project.projectId, userId).transformWith {
+            case Success(configuration) => Future.successful(project, Some(configuration))
+            case Failure(NotFound(_))   => Future.successful(project, None)
+            case _                      => internalError("fetch configuration")
+          }
+
+        for {
+          projects <- eventualProjects
+          projWithConf <- Future.sequence(projects.map(getProjectWithConfig))
+        } yield projWithConf.collect {
+          case (proj, Some(_)) if hasConfig => Some(proj)
+          case (proj, None) if !hasConfig   => Some(proj)
+          case _                            => None
+        }.flatten
```

Comment:
`case _ =>` in `collect` function usually indicates that you are doing something wrong. It doesn't make sense to collect data with total function, with total functions we use `map` / `flatMap`
```suggestion
        } yield projWithConf.collect {
          case (proj, Some(_)) if hasConfig => proj
          case (proj, None) if !hasConfig   => proj
        }
```

Reply:
Removed

Reply:
Your changes do not contain everything that was suggested here